Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
++                    //Paper end
+                     chunkcoordintpair1 = new ChunkCoordIntPair(k1, l1);
+                     flag5 = false;
+                     flag6 = true;
+@@ -1365,8 +1392,12 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+     public Stream<EntityPlayer> a(ChunkCoordIntPair chunkcoordintpair, boolean flag) {
+         return this.playerMap.a(chunkcoordintpair.pair()).filter((entityplayer) -> {
+             int i = b(chunkcoordintpair, entityplayer, true);
+-
+-            return i > this.viewDistance ? false : !flag || i == this.viewDistance;
++            //Paper start ticket level pp-view-dist
++            PlayerChunk playerchunk = this.getVisibleChunk(chunkcoordintpair.pair());
++            //check ticking chunk
++            if(playerchunk == null || playerchunk.getTickingChunk() == null) return false;
++            return i > entityplayer.getTrackViewDistance() ? false : !flag || i == entityplayer.getTrackViewDistance();
++            //Paper end
+         });
+     }
+ 
+@@ -1411,7 +1442,7 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+         org.spigotmc.AsyncCatcher.catchOp("entity untrack"); // Spigot
+         if (entity instanceof EntityPlayer) {
+             EntityPlayer entityplayer = (EntityPlayer) entity;
+-
++            this.chunkDistanceManager.fullRemove((EntityPlayer) entity); //Paper ticket level pp-view-dist
+             this.a(entityplayer, false);
+             ObjectIterator objectiterator = this.trackedEntities.values().iterator();
+ 
+@@ -1616,19 +1647,20 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+             org.spigotmc.AsyncCatcher.catchOp("player tracker update"); // Spigot
+             if (entityplayer != this.tracker) {
+                 Vec3D vec3d = (new Vec3D(entityplayer.locX, entityplayer.locY, entityplayer.locZ)).d(this.tracker.getPositionVector()); // MC-155077, SPIGOT-5113
+-                int i = Math.min(this.trackingDistance, (PlayerChunkMap.this.viewDistance - 1) * 16);
++                //Paper start ticket level pp-view-dist
++                int i = Math.min(this.trackingDistance, (entityplayer.getTrackViewDistance() - 1) * 16);
+                 boolean flag = vec3d.x >= (double) (-i) && vec3d.x <= (double) i && vec3d.z >= (double) (-i) && vec3d.z <= (double) i && this.tracker.a(entityplayer);
++                ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(this.tracker.chunkX, this.tracker.chunkZ);
++                PlayerChunk playerchunk = PlayerChunkMap.this.getVisibleChunk(chunkcoordintpair.pair());
++                //untrack if chunk is not entity ticking,
++                //as the default tracking range check - 1 only operates correctly when all view distances are equal and constant
+ 
+-                if (flag) {
++                if (flag && playerchunk != null && (playerchunk.getEntityTickingChunk() != null)) {
```

Comment:
I had misread mojang's code (neighbours require 1 radius of loaded chunks, not 2...) so this technically should be the ticking chunk

Reply:
solved in rewrite already